internal/provider/resource_dns_cname_record.go

```diff
@@ -1,137 +1,337 @@
-				Type:         schema.TypeString,
-				Required:     true,
-				ValidateFunc: validateZone,
+			"cname": schema.StringAttribute{
+				Required: true,
+				Validators: []validator.String{
+					dnsvalidator.IsZoneNameValid(),
+				},
+				Description: "The canonical name this record will point to.",
 			},
-			"ttl": {
-				Type:     schema.TypeInt,
+			"ttl": schema.Int64Attribute{
 				Optional: true,
-				ForceNew: true,
-				Default:  3600,
+				Computed: true,
+				Default:  int64default.StaticInt64(3600),
+				PlanModifiers: []planmodifier.Int64{
+					int64planmodifier.RequiresReplace(),
+				},
+				Description: "The TTL of the record set. Defaults to `3600`.",
+			},
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "Always set to the fully qualified domain name of the record.",
 			},
 		},
 	}
 }
 
-func resourceDnsCnameRecordCreate(d *schema.ResourceData, meta interface{}) error {
+func (d *dnsCNAMERecordResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
 
-	d.SetId(resourceFQDN(d))
+	client, ok := req.ProviderData.(*DNSClient)
 
-	return resourceDnsCnameRecordUpdate(d, meta)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *DNSClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	d.client = client
 }
 
-func resourceDnsCnameRecordRead(d *schema.ResourceData, meta interface{}) error {
+func (d *dnsCNAMERecordResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan cnameRecordResourceModel
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	config := dnsConfig{
+		Name: plan.Name.ValueString(),
+		Zone: plan.Zone.ValueString(),
+	}
+	rec_fqdn := resourceFQDN_framework(config)
+	plan.ID = types.StringValue(rec_fqdn)
+
+	msg := new(dns.Msg)
+	msg.SetUpdate(plan.Zone.ValueString())
 
-	answers, err := resourceDnsRead(d, meta, dns.TypeCNAME)
+	rrStrInsert := fmt.Sprintf("%s %d CNAME %s", rec_fqdn, plan.TTL.ValueInt64(), plan.CNAME.ValueString())
+
+	rr_insert, err := dns.NewRR(rrStrInsert)
 	if err != nil {
-		return err
+		resp.Diagnostics.AddError(fmt.Sprintf("Error reading DNS record (%s):", rrStrInsert), err.Error())
+		return
 	}
+	msg.Insert([]dns.RR{rr_insert})
 
-	if len(answers) > 0 {
+	r, err := exchange_framework(msg, true, d.client)
+	if err != nil {
+		resp.State.RemoveResource(ctx)
+		resp.Diagnostics.AddError("Error updating DNS record:", err.Error())
+		return
+	}
+	if r.Rcode != dns.RcodeSuccess {
+		resp.State.RemoveResource(ctx)
+		resp.Diagnostics.AddError(fmt.Sprintf("Error updating DNS record: %v", r.Rcode), dns.RcodeToString[r.Rcode])
+		return
+	}
+
+	answers, diags := resourceDnsRead_framework(config, d.client, dns.TypeCNAME)
+	if diags != nil {
```

Comment:
Please note that checking `diags != nil` like this here and in other locations could cause logic issues if:

- The diagnostics is an empty slice (instantiated with 0 elements)
- The diagnostics contain only warnings

Generally its preferable to always call `Append()` (it handles `nil` and 0 elements fine) and do logic only when returned diagnostics has errors, e.g. `if XXX.HasError()`